lib/stm32/common/st_usbfs_core.c

```diff
@@ -97,27 +109,51 @@ void st_usbfs_ep_setup(usbd_device *dev, uint8_t addr, uint8_t type,
 	USB_SET_EP_TYPE(addr, typelookup[type]);
 
 	if (dir || (addr == 0)) {
-		USB_SET_EP_TX_ADDR(addr, dev->pm_top);
+		if (type == USB_ENDPOINT_ATTR_ISOCHRONOUS) {
+			/* isochronous endpoints always use double buffering
+			 * and are unidirectional */
+			USB_SET_EP_RX_STAT(addr, USB_EP_RX_STAT_DISABLED);
+			USB_SET_EP_TX_0_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_0_COUNT(addr, max_size);
+			dev->pm_top += max_size;
+			USB_SET_EP_TX_1_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_1_COUNT(addr, max_size);
+		} else {
+			USB_SET_EP_TX_ADDR(addr, dev->pm_top);
+			USB_SET_EP_TX_COUNT(addr, max_size);
+		}
+		dev->pm_top += max_size;
 		if (callback) {
 			dev->user_callback_ctr[addr][USB_TRANSACTION_IN] =
 			    (void *)callback;
 		}
 		USB_CLR_EP_TX_DTOG(addr);
-		USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_NAK);
-		dev->pm_top += max_size;
+		if (type == USB_ENDPOINT_ATTR_ISOCHRONOUS)
+			USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_VALID); // or DISABLED?
+		else
+			USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_NAK);
 	}
 
 	if (!dir) {
-		uint16_t realsize;
-		USB_SET_EP_RX_ADDR(addr, dev->pm_top);
-		realsize = st_usbfs_set_ep_rx_bufsize(dev, addr, max_size);
+		uint16_t real_size;
+		if (type == USB_ENDPOINT_ATTR_ISOCHRONOUS) {
+			USB_SET_EP_TX_STAT(addr, USB_EP_TX_STAT_DISABLED);
+			USB_SET_EP_RX_0_ADDR(addr, dev->pm_top);
+			USB_SET_EP_RX_0_COUNT(addr, convert_bufsize(max_size, &real_size) << 10);
+			dev->pm_top += real_size;
+			USB_SET_EP_RX_1_ADDR(addr, dev->pm_top);
+			USB_SET_EP_RX_1_COUNT(addr, convert_bufsize(max_size, &real_size) << 10);
```

Comment:
I find the idea that you use this _COUNT() macro in _completely different_ manners depending on the direction to be extremely confusing.  I realise this is "behind the scenes" code that isn't user facing, but why drop the _set_ep_rx_bufsize() usage?  if you've dropped it, is it still useful at all? if it's no longer useful, why still keep it?

Reply:
It is the register layout that makes this confusing. The RX count is encoded and sits at bits 10 and up, whereas the TX count is plain, written at bits 0 and up. The question is if the bit shifting <<10 should be part of the _SET_EP_RX macros.

Reply:
But then if we "hide" the shifting should we hide the encoding too?

Reply:
> why still keep it

I thought it was exposed for applications, since it wasn't "static". Would be good to get rid of otherwise. It would make this function more difficult to read if we would use _set_ep_rx_bufsize in one case and the direct code in the other case. The symmetry makes it easier to read.